File: src/org/simbrain/workspace/serialization/ArchivedCoupling.java
package org.simbrain.workspace.serialization;

import org.simbrain.workspace.*;

/**
 * Class used to represent a coupling in the archive.
 *
 * @author Matt Watson
 */
class ArchivedCoupling {

    /**
     * The source attribute for the coupling.
     */
    private ArchivedAttribute producer;

    /**
     * The target attribute for the coupling.
     */
    private ArchivedAttribute consumer;

    /**
     * Creates a new instance.
     *
     * @param producer The producer attribute.
     * @param consumer The consumer attribute.
     */
    ArchivedCoupling(ArchivedAttribute producer, ArchivedAttribute consumer) {
        this.producer = producer;
        this.consumer = consumer;
    }

    public ArchivedAttribute getProducer() {
        return producer;
    }

    public ArchivedAttribute getConsumer() {
        return consumer;
    }

    public Producer createProducer(Workspace workspace) {
        AttributeContainer object = getObjectFromWorkspace(workspace, producer);
        String method = producer.getMethodName();
        return CouplingUtils.getProducer(object, method);
    }

    public Consumer createConsumer(Workspace workspace) {
        AttributeContainer container = getObjectFromWorkspace(workspace, consumer);
        String method = consumer.getMethodName();
        return CouplingUtils.getConsumer(container, method);
    }

    private AttributeContainer getObjectFromWorkspace(Workspace workspace, ArchivedAttribute attribute) {
        WorkspaceComponent component = workspace.getComponent(attribute.getComponentId());
        AttributeContainer container = component.getObjectFromKey(attribute.getId());
        if (container == null) {
            throw new RuntimeException(String.format("Failed to retrieve object %s from serialized component %s.", attribute.getId(), attribute.getComponentId()));
        }
        return container;
    }

}

File: src/main/kotlin/org/simbrain/workspace/serialization/ArchivedCoupling.kt
/**
 * Archive form of a [org.simbrain.workspace.couplings.Coupling]: the producer and consumer as
 * [ArchivedAttribute]s plus the coupling's transform chain, which is serialized as the operation objects
 * themselves. The transforms field is nullable because archives written before transforms existed have
 * no such element; restore treats null as an empty chain.
 */
package org.simbrain.workspace.serialization

import org.simbrain.workspace.AttributeContainer
import org.simbrain.workspace.Consumer
import org.simbrain.workspace.Producer
import org.simbrain.workspace.Workspace
import org.simbrain.workspace.couplings.CouplingOperation

class ArchivedCoupling(
    val producer: ArchivedAttribute,
    val consumer: ArchivedAttribute,
    val transforms: List<CouplingOperation<*, *>>? = null
) {

    fun createProducer(workspace: Workspace): Producer = with(workspace.couplingManager) {
        getObjectFromWorkspace(workspace, producer).getProducer(producer.methodName)
    }

    fun createConsumer(workspace: Workspace): Consumer = with(workspace.couplingManager) {
        getObjectFromWorkspace(workspace, consumer).getConsumer(consumer.methodName)
    }

    /**
     * Find the attribute container corresponding to an archived attribute object.
     */
    private fun getObjectFromWorkspace(workspace: Workspace, attribute: ArchivedAttribute): AttributeContainer {
        val component = workspace.getComponent(attribute.componentId)
            ?: throw RuntimeException("No component ${attribute.componentId} found for archived attribute ${attribute.attributeId}.")
        return component.attributeContainers.find { it.id == attribute.attributeId }
            ?: throw RuntimeException(
                "Failed to retrieve object ${attribute.attributeId} from serialized component ${attribute.componentId}."
            )
    }
}
